File: cli-rust/crates/oa-cmd-auth/src/oauth_env.rs
/// OAuth environment detection (remote/VPS vs local).
///
/// Detects whether the current environment is remote (SSH, containers,
/// headless Linux) so that OAuth flows can adapt to show manual URL
/// instructions instead of attempting to open a local browser.
///
/// Source: `src/commands/oauth-env.ts`

/// Check whether the current process is running in a remote/VPS environment.
///
/// Returns `true` when any of the following conditions are met:
/// - SSH session indicators (`SSH_CLIENT`, `SSH_TTY`, `SSH_CONNECTION`)
/// - Container indicators (`REMOTE_CONTAINERS`, `CODESPACES`)
/// - Headless Linux (no `DISPLAY` or `WAYLAND_DISPLAY`) that is not WSL
///
/// Source: `src/commands/oauth-env.ts` - `isRemoteEnvironment`
#[must_use]
pub fn is_remote_environment() -> bool {
    // SSH session detection
    if std::env::var("SSH_CLIENT").is_ok()
        || std::env::var("SSH_TTY").is_ok()
        || std::env::var("SSH_CONNECTION").is_ok()
    {
        return true;
    }

    // Container / cloud IDE detection
    if std::env::var("REMOTE_CONTAINERS").is_ok() || std::env::var("CODESPACES").is_ok() {
        return true;
    }

    // Headless Linux (no display server, not WSL)
    if cfg!(target_os = "linux")
        && std::env::var("DISPLAY").is_err()
        && std::env::var("WAYLAND_DISPLAY").is_err()
        && !is_wsl()
    {
        return true;
    }

    false
}

/// Check if we are running in Windows Subsystem for Linux.
///
/// Source: `src/infra/wsl.ts` - `isWSLEnv`
fn is_wsl() -> bool {
    if !cfg!(target_os = "linux") {
        return false;
    }

    // Check WSL-specific environment variables
    if std::env::var("WSL_DISTRO_NAME").is_ok() || std::env::var("WSL_INTEROP").is_ok() {
        return true;
    }

    // Check /proc/version for WSL indicator
    std::fs::read_to_string("/proc/version")
        .ok()
        .map_or(false, |content| {
            let lower = content.to_lowercase();
            lower.contains("microsoft") || lower.contains("wsl")
        })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn is_remote_returns_bool() {
        // We cannot control the environment in unit tests, but we can at
        // least verify the function runs without panicking.
        let _result = is_remote_environment();
    }

    #[test]
    fn is_wsl_returns_bool() {
        let _result = is_wsl();
    }

    #[test]
    fn remote_detection_does_not_panic() {
        // Ensure the function handles missing env vars gracefully
        let result = is_remote_environment();
        assert!(result || !result); // always true, just ensuring no panic
    }
}

File: backend/internal/goproviders/onboard/oauth_env.go
// onboard/oauth_env.go — OAuth 环境检测
// 对应 TS 文件: src/commands/oauth-env.ts
// 检测是否在远程环境（SSH、Codespaces、无显示的 Linux）中运行。
package onboard

import (
	"os"
	"runtime"
)

// IsRemoteEnvironment 检查是否在远程环境中运行。
// 对应 TS: isRemoteEnvironment()
func IsRemoteEnvironment() bool {
	// SSH 连接
	if os.Getenv("SSH_CLIENT") != "" || os.Getenv("SSH_TTY") != "" || os.Getenv("SSH_CONNECTION") != "" {
		return true
	}

	// 容器/Codespaces 环境
	if os.Getenv("REMOTE_CONTAINERS") != "" || os.Getenv("CODESPACES") != "" {
		return true
	}

	// Linux 无图形显示（排除 WSL）
	if runtime.GOOS == "linux" &&
		os.Getenv("DISPLAY") == "" &&
		os.Getenv("WAYLAND_DISPLAY") == "" &&
		!isWSLEnv() {
		return true
	}

	return false
}

// isWSLEnv 检测是否在 WSL 环境中。
func isWSLEnv() bool {
	// 检查 WSL 特有的环境变量
	if os.Getenv("WSL_DISTRO_NAME") != "" || os.Getenv("WSLENV") != "" {
		return true
	}
	// 检查 /proc/version 中是否包含 WSL 标志
	data, err := os.ReadFile("/proc/version")
	if err != nil {
		return false
	}
	content := string(data)
	return contains(content, "microsoft") || contains(content, "WSL")
}

// contains 检查字符串是否包含子串（不区分大小写用不到，此处精确匹配）。
func contains(s, substr string) bool {
	return len(s) >= len(substr) && searchString(s, substr)
}

func searchString(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
